storage-proofs/src/piece_inclusion_proof.rs

```diff
@@ -1,255 +1,473 @@
+}
+
+pub fn piece_inclusion_proofs<H: Hasher>(
+    piece_specs: &[PieceSpec],
+    tree: &MerkleTree<H::Domain, H::Function>,
+) -> Result<Vec<PieceInclusionProof<H>>> {
+    piece_specs
         .iter()
-        .zip(pieces)
-        .all(|(proof, piece)| proof.verify(root, piece))
+        .map(|piece_spec| PieceInclusionProof::new((*piece_spec).clone(), tree))
+        .collect()
+}
+
+impl<H: Hasher> PieceInclusionProof<H> {
+    pub fn new(
+        piece_spec: PieceSpec,
+        tree: &MerkleTree<H::Domain, H::Function>,
+    ) -> Result<PieceInclusionProof<H>> {
+        let PieceSpec {
+            comm_p,
+            position: first_leaf,
+            length: leaf_count,
+        } = piece_spec;
+
+        let last_leaf = first_leaf + (leaf_count - 1);
+
+        // For now, we only handled aligned pieces.
+        if !piece_spec.is_aligned(tree.len()) {
+            return Err(Error::UnalignedPiece);
+        }
+
+        let first_proof = tree.gen_proof(first_leaf);
+        let last_proof = tree.gen_proof(last_leaf);
+
+        // Including first leaf (item).
+        let proof_length = piece_spec.height() + 1;
+
+        // Find the first common hash (at same position) in first and last proof,
+        // then discard all previous hashes and path bits, and create a new proof containing
+        // only the common part.
+        //
+        // For an aligned piece, there is guaranteed to be a common element.
+        // That is, the merkle root of the piece must be a node of the tree (and therefore in the paths
+        // of all leaves of the piece (including the first and last).
+        for (i, (a, b)) in first_proof
+            .lemma()
+            .iter()
+            .zip(last_proof.lemma().iter())
+            .enumerate()
+        {
+            if (a == b) && (i == proof_length) {
+                let proof_elements = if i > first_proof.path().len() {
+                    vec![]
+                } else {
+                    let mut lemma = first_proof.lemma()[i..].to_vec();
+                    lemma.insert(0, H::Domain::try_from_bytes(&comm_p)?);
+                    let path = first_proof.path()[i - 1..].to_vec();
+
+                    let p = Proof::<H::Domain>::new(lemma, path);
+
+                    p.lemma()[1..p.lemma().len() - 1].to_vec()
+                };
+
+                return Ok(PieceInclusionProof {
+                    proof_elements,
+                    position: first_leaf,
+                    piece_leaves: leaf_count,
+                });
+            }
+        }
+
+        // There must be a common root, if only the root of the entire tree.
+        Err(Error::MalformedMerkleTree)
+    }
+
+    fn leaf_path(leaf: usize, tree_height: usize) -> Vec<bool> {
+        let mut height = tree_height;
+        let mut rising_leaf = leaf;
+        let mut path = Vec::new();
+
+        while height > 0 {
+            // If this leaf's 1-bit is 1, then it is a right branch, so `is_right` is true.
```

Comment:
`0-bit is 1` I believe

Reply:
I meant 1-bit as in one's place (as opposed to 2-bit, 4-bit, etc.). Rewording to avoid the issue.